### PR TITLE
Adding a new framework for concurrency testing in geode

### PR DESCRIPTION
**Pull request by upthewaterspout**
Added a new junit runner, ConcurrentTestRunner, for running a test that
has parallel threads.

The runner currently runs the test using Java PathFinder, which will
run the test with all interleavings of the threads.

Added an example test to geode-core for the FilterProfile serialization
logic.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ConcurrentTestRunner.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+ * </code>
+ *
+ * ConcurrentTestRunner currently executes tests using Java Pathfinder, which will run the test with
+ * all thread interleavings.
+ */
+public class ConcurrentTestRunner extends ParentRunner<FrameworkMethod> {
+  /**
+   * Delegate to actually run the test.
+   */
+  private Runner runner = new JpfRunner();;
```
Comment (pivotal-jbarrett) [inline geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ConcurrentTestRunner.java]: ;;?

File: geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ParallelExecutor.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+   * Add a task to run in parallel
+   */
+  default <T> Future<T> inParallel(RunnableWithException runnable) {
+    return inParallel(runnable, null);
+  }
+
+  /**
+   * Add a task to run in parallel
+   */
+  default <T> Future<T> inParallel(RunnableWithException runnable, T value) {
```
Comment (pivotal-jbarrett) [inline geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ParallelExecutor.java]: What is the use case for passing in a value that is just echoed back via the future?

File: geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/RunnableWithException.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * the License.
+ */
+package org.apache.geode.test.concurrency;
+
+import java.io.Serializable;
+
+/**
+ * Like Runnable, but allows the run method to throw an exception
+ */
+public interface RunnableWithException extends Serializable {
```
Comment (pivotal-jbarrett) [inline geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/RunnableWithException.java]: What does this interface gain us that `Callable` doesn't cover?

File: geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import org.apache.geode.test.concurrency.ConcurrentTestRunner;
+import org.apache.geode.test.concurrency.ParallelExecutor;
+
+@RunWith(ConcurrentTestRunner.class)
+public class FilterProfileConcurrencyTest {
+
+  @Test
+  public void serializationOfFilterProfileWithConcurrentUpdateShouldSucceed(
+      ParallelExecutor executor) throws Exception {
+    warmUp();
```
Comment (pivotal-jbarrett) [inline geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java]: Would it make sense to bake warmup into the test framework. Since these should be unit test and good unit tests have no side effects and run in a very short time then wouldn't it be okay to run the test twice, once for warmup and the second inside the JPF context?

File: geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ParallelExecutor.java (hunk 2 above)
Comment (PivotalSarge) [reply geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ParallelExecutor.java]: Is it so individual runs can be identified?

File: geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ConcurrentTestRunner.java (hunk 1 above)
Comment (gesterzhou) [reply geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ConcurrentTestRunner.java]: I think we should have a regular ConcurrentTestRunner, then extends or implements one class to use JpfRunner. Not to assume to use JpfRunner as default.

File: geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java (hunk 4 above)
Comment (galen-pivotal) [reply geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java]: My intuition tells me it would be nice to have an escape hatch in case we were testing issues that had to do with setting up a Cache or classloading or logging or something.

Comment (pivotal-jbarrett) [reply geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java]: @galen-pivotal good point. So maybe an annotation to `@warmup` or `@nowarmup` depending on the default?

File: geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ParallelExecutor.java (hunk 2 above)
Comment (upthewaterspout) [reply geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ParallelExecutor.java]: Maybe not too much. I was copying a similar pattern from the ExecutorService API. I can remove this method until we actually need it.

File: geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/RunnableWithException.java (hunk 3 above)
Comment (upthewaterspout) [reply geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/RunnableWithException.java]: It means you don't have to return a value. So you can do

executor.inParallel(() -> System.out.println("hi"))

instead of 

executor.inParallel(() -> { 
  System.out.println("hi");
  return null;
)

File: geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java (hunk 4 above)
Comment (upthewaterspout) [reply geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java]: I think you are right - the test shouldn't have to do this warmup business. Let me see if I can work something in (with a configuration option to disable it).

File: geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ConcurrentTestRunner.java (hunk 1 above)
Comment (upthewaterspout) [reply geode-concurrency-test/src/main/java/org/apache/geode/test/concurrency/ConcurrentTestRunner.java]: @gesterzhou - I updated the PR. You can now configure which runner to use.

File: geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java (hunk 4 above)
Comment (upthewaterspout) [reply geode-core/src/jpfTest/java/org/apache/geode/internal/cache/FilterProfileConcurrencyTest.java]: After more investigation, most of the issues with classloading were actually due to initializing the logger - it updates a whole bunch of shared state. The next biggest issue was InternalDataSerializer static initialization - that updates a bunch of shared state as well.

I completely override the logger classes to just return my own logger that does nothing. I also set the jpf flags that treat static initializers and final initialization as atomic. I added configuration options so that the user can pass in properties configuring JPF, in case they want to change that behavior.

Long story short, the test no longer needs to warm up, and it runs a bit faster than it did before.
